Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java

```diff
@@ -327,138 +351,176 @@ private int deletePoint(short nodeOffset, double[] point, int level) {
-        } else {
-            return addPoint(rootIndex, pointValue, pointIndex);
-        }
-    }
-
     /**
-     * This function adds a point to the tree recursively. The algorithm for adding
-     * a point is as follows:
-     * <ol>
-     * <li>At the current node we create a new bounding box by merging the point
-     * with the existing box.</li>
-     * <li>We pick a dimension and a value of cut.</li>
-     * <li>If the cut falls outside the existing box we create a new node and
-     * replace the current node with the new node. We then add the current node as
-     * the child node to the new node. We create another leaf node containing the
-     * point that we want to insert and add it as the other child of the node that
-     * we have created earlier.</li>
-     * <li>If the cut falls inside the existing box. We follow the cut of the
-     * existing box and move to right or left child of the current node based on the
-     * existing cut.</li>
-     * </ol>
+     * The following class corresponds to the state stored during the addPoint
+     * operation; the sibling (after the addition) of the point being added, the cut
+     * dimension, the cut value and the bounding box. Saving the same bounding box
+     * as is used to generate the random cut appears to be necessary from the
+     * perspective of precision/loss there of.
      *
-     * @param nodeOffset the current node in the tree we are on
-     * @param point      the point that we want to add to the tree
-     * @param pointIndex is the location of the point in pointstore
+     * The assumption is that the process would begin at the leaf and bubble up.
+     *
+     * The checkContainsAndUpdateCut method provides a mechanism to update the state
+     * as we visit parents recursively, in the current context, nodeOffset.
+     *
+     * The effectChange methiod effects the change, where it takes the additional
+     * information of the parent where the addition of the point no longer updates
+     * any bounding box.
      */
 
-    private int addPoint(short nodeOffset, double[] point, int pointIndex) {
-        if (isLeaf(nodeOffset) && pointStore.pointEquals(leafNodes.pointIndex[nodeOffset - maxSize], point)) {
-            // the inserted point is equal to an existing leaf point
-            ++leafNodes.mass[nodeOffset - maxSize];
-            return leafNodes.pointIndex[nodeOffset - maxSize];
+    class CandidateChange {
+        short siblingOffset;
+        int cutDimension;
+        double cutValue;
+        BoundingBox savedBox;
+
+        void setCandidateChange(short sibling, short dim, double val, BoundingBox box) {
+            siblingOffset = sibling;
+            cutDimension = dim;
+            cutValue = val;
+            savedBox = box;
         }
 
-        // either the node is not a leaf, or else it's a leaf node containing a
-        // different point
+        void effectChange(int pointIndex, double[] point, int parentIndex) {
```

Comment:
We could call this method `resolve` to match the flag name.

Reply:
Fixed